openest/generate/calculation.py

```diff
@@ -86,8 +86,34 @@ def describe():
         """
         raise NotImplementedError()
 
+class RecursiveCalculation(Calculation):
+    """Calculation that contains subcalculations.
 
-class FunctionalCalculation(Calculation):
+    This is a light-weight ABC mainly for passing messages down the tree.
```

Comment:
Is that how the term abstract class are used in python-- as interfaces? We used to (like in C++) call those pure virtual classes, but then an abstract classes was still anything that had any abstract methods and therefore couldn't be instantiated.

Reply:
Yeah, you're totally right and this is an interesting design discussion. In that abstract classes can contain concrete methods  and can't be implemented. Concrete classes contain only concrete methods.

[ABCs](https://docs.python.org/3/glossary.html#term-abstract-base-class) are explicitly an interface thing in Python. They're a pythony version of traditional OOP virtual classes.

Generally when we're dealing with class inheritance a python best practice is to distinguish *why* you're doing inheritance. You want to separate interface inheritance from implementation inheritance. Interface inheritance comes form ABCs and implementation comes from *one* other concrete class -- multiple mixins are okay. You can also inherit multiple interfaces (ABCs).

So, yeah, it's an abstract class. It's not really an ABC. If we wanted them to be ABCs... we'd need to be sure that (for e.g.) anything inheriting that ABC keeps the same signature of overloaded methods - which I don't think we do. It feels misleading to call these ABCs because we violate the [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle)  so often for objects that we claim share the same interface. In practice, ABCs don't really represent the intent of these abstract classes.

Reply:
I see also that we shouldn't be calling any of our classes (even pure interfaces) ABCs, since we're not using the `abc`. By the way, although there are cases where I've changed the signature of an overloaded method, I considered this an undesirable expedient-- I needed to expand things and didn't want to go back at the time. But I've been trying to clear these up as I come across them.